src/server/scripts/Spells/spell_warlock.cpp

```diff
@@ -1216,6 +1219,40 @@ class spell_warl_siphon_life : public AuraScript
     }
 };
 
+// 6353 - Soul Fire
+class spell_warl_soul_fire : public SpellScript
+{
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellInfo ({ SPELL_WARLOCK_SOUL_FIRE_ENERGIZE, SPELL_WARLOCK_WITHER, SPELL_WARLOCK_WITHER_PERIODIC_DAMAGE, SPELL_WARLOCK_IMMOLATE_PERIODIC });
+    }
+
+    void HandleAfterCast(SpellEffIndex /*effIndex*/) const
+    {
+        Unit* caster = GetCaster();
+
+        caster->CastSpell(caster, SPELL_WARLOCK_SOUL_FIRE_ENERGIZE, CastSpellExtraArgsInit{
+            .TriggerFlags = TRIGGERED_IGNORE_CAST_IN_PROGRESS | TRIGGERED_DONT_REPORT_CAST_ERROR,
+            .TriggeringSpell = GetSpell()
+        });
+    }
+
+    void HandleSoulFire() const
+    {
+        GetCaster()->CastSpell(GetHitUnit(), GetCaster()->HasAura(SPELL_WARLOCK_WITHER) ? SPELL_WARLOCK_WITHER_PERIODIC_DAMAGE : SPELL_WARLOCK_IMMOLATE_PERIODIC,
```

Comment:
take this with a grain of salt, but from a readability and maintainability perspective (and debuggability), it's always best to encapsulate "complex" expressions like ternary operators in a variable and passing it